### PR TITLE
Reconnect in sbp_rx if we get disconnected [ESD-968]

### PR DESCRIPTION
**Pull request by silverjam**
## Design Notes

In `pk_endpoint_t` we disconnect clients that fall behind and failed to empty their buffers:
  - https://github.com/swift-nav/piksi_buildroot/blob/v2.2.0-release/package/libpiksi/libpiksi/src/endpoint.c#L628

This means that clients that wish to continue receiving data must notice the disconnect and manually reconnect in order to receive data again.  This is presumably something that was handled "under the covers" by nanomsg/zmq.

## Testing

Tested with `PoseDaemon` and confirmed that initially on start-up it gets behind on servicing data causing it's connection to the `endpoint_router` to be dropped.  Once it reconnects and behind processing data will remain "ahead" and no further reconnects are necessary.

### CODE REVIEW

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v2.2.9-2-gd1133413
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=815867)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.2.9-2-gd1133413&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.2.9-2-gd1133413&groupby_key=firmware&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v2.2.9-2-gd1133413
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.2.9-2-gd1133413)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.2.9-2-gd1133413/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.2.9-2-gd1133413

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v2.2.9-5-gec2857e6
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=815958)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.2.9-5-gec2857e6&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.2.9-5-gec2857e6&groupby_key=firmware&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v2.2.9-5-gec2857e6
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.2.9-5-gec2857e6)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.2.9-5-gec2857e6/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.2.9-5-gec2857e6

**Comment by silverjam** [comment]
Crash is weird, looks like we're hitting this assert: https://github.com/nodejs/node/blob/master/deps/uv/src/unix/linux-core.c#L244

Based on this disassembly:
```
   0xb6fa9520 <+344>:	mov	r1, r9
   0xb6fa9524 <+348>:	vstr	d16, [r7, #-136]	; 0xffffff78
   0xb6fa9528 <+352>:	bl	0xb6fa7690 <uv__epoll_ctl@plt>
   0xb6fa952c <+356>:	cmp	r0, #0
   0xb6fa9530 <+360>:	beq	0xb6fa9590 <uv__io_poll.part.5+456>
   0xb6fa9534 <+364>:	bl	0xb6fa7c9c <__errno_location@plt>
   0xb6fa9538 <+368>:	ldr	r3, [r0]
   0xb6fa953c <+372>:	cmp	r3, #17
   0xb6fa9540 <+376>:	beq	0xb6fa9548 <uv__io_poll.part.5+384>
   0xb6fa9544 <+380>:	bl	0xb6fa8260 <abort@plt>
=> 0xb6fa9548 <+384>:	cmp	r9, #1
   0xb6fa954c <+388>:	beq	0xb6fa9574 <uv__io_poll.part.5+428>
   0xb6fa9550 <+392>:	ldr	r3, [pc, #1240]	; 0xb6fa9a30 <uv__io_poll.part.5+1640>
   0xb6fa9554 <+396>:	mov	r2, #247	; 0xf7
   0xb6fa9558 <+400>:	ldr	r1, [pc, #1236]	; 0xb6fa9a34 <uv__io_poll.part.5+1644>
   0xb6fa955c <+404>:	ldr	r0, [pc, #1236]	; 0xb6fa9a38 <uv__io_poll.part.5+1648>
   0xb6fa9560 <+408>:	add	r3, pc, r3
   0xb6fa9564 <+412>:	add	r3, r3, #45	; 0x2d
   0xb6fa9568 <+416>:	add	r1, pc, r1
   0xb6fa956c <+420>:	add	r0, pc, r0
   0xb6fa9570 <+424>:	b	0xb6fa9498 <uv__io_poll.part.5+208>
```

And this backtrace:
```
(gdb) bt
#0  0xb6dd0e70 in raise () from /piksi_buildroot/buildroot/output/host/arm-buildroot-linux-gnueabihf/sysroot/lib/libc.so.6
#1  0xb6dd2134 in abort () from /piksi_buildroot/buildroot/output/host/arm-buildroot-linux-gnueabihf/sysroot/lib/libc.so.6
#2  0xb6fa9548 in uv__io_poll.part.5 (loop=0xb6ff94c0, timeout=-1225912104) at src/unix/linux-core.c:245
#3  0xb6fb06d4 in uv_run (loop=0x2a240, mode=UV_RUN_DEFAULT) at src/unix/core.c:368
#4  0xb6f75d28 in __fn__.8781 ()
   from /piksi_buildroot/buildroot/output/host/arm-buildroot-linux-gnueabihf/sysroot/usr/lib/libpiksi.so
#5  0x00000000 in ?? ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
```

Crash only happens if I set the `resource_monitor` to trigger at 1s interval.  Will need to investigate more tomorrow.

File: package/libpiksi/libpiksi/src/sbp_rx.c [hunk 1]
```diff
@@ -105,8 +143,14 @@ int sbp_rx_attach(sbp_rx_ctx_t *ctx, pk_loop_t *pk_loop)
   assert(ctx != NULL);
   assert(pk_loop != NULL);
 
+  // Must assign loop anew or assign the same loop
```
Comment (pmiettinen) [inline package/libpiksi/libpiksi/src/sbp_rx.c]: comment style

Comment (silverjam) [reply package/libpiksi/libpiksi/src/sbp_rx.c]: fixed, realized I didn't need the new structure member

**Comment by silverjam** [comment]
The issues with `resource_monitor` was apparently a bug/weirdness in libuv, apparently there's a race if the same fd gets re-created which was just closed?

References:
- https://github.com/libuv/libuv/issues/1495
- https://github.com/nodejs/node-v0.x-archive/issues/4558

Introducing a "fake fd" to force it to allocate a different fd seems to make this work... however now this reveals that `resource_monitor` drops a lot of data if it's collection interval is set too low (e.g. 1s):

```
Dec 18 08:15:55 piksi local0.warn resource_monitor[2644]: sbp_rx: reader disconnected, reconnecting...
Dec 18 08:15:59 piksi local0.warn router[987]: send_impl: sendmsg returned EAGAIN, disconnecting and dropping 163865 queued bytes (path: ipc:///var/run/sockets/internal.pub, node: 0x4d400) (endpoint.c:633)
Dec 18 08:15:59 piksi local0.warn resource_monitor[2644]: process_ss_entry: unknown socket type: raw (query_sockets.c:273)
Dec 18 08:15:59 piksi local0.warn resource_monitor[2644]: sbp_rx: reader disconnected, reconnecting...
Dec 18 08:16:02 piksi local0.warn router[987]: send_impl: sendmsg returned EAGAIN, disconnecting and dropping 164057 queued bytes (path: ipc:///var/run/sockets/internal.pub, node: 0x4d400) (endpoint.c:633)
Dec 18 08:16:02 piksi local0.warn resource_monitor[2644]: sbp_rx: reader disconnected, reconnecting...
Dec 18 08:16:04 piksi local0.warn router[987]: send_impl: sendmsg returned EAGAIN, disconnecting and dropping 164008 queued bytes (path: ipc:///var/run/sockets/internal.pub, node: 0x4d400) (endpoint.c:633)
Dec 18 08:16:05 piksi local0.warn resource_monitor[2644]: sbp_rx: reader disconnected, reconnecting...
Dec 18 08:16:06 piksi authpriv.info dropbear[15203]: Child connection from 192.168.22.107:55127
Dec 18 08:16:07 piksi authpriv.notice dropbear[15203]: Auth succeeded with blank password for 'root' from 192.168.22.107:55127
Dec 18 08:16:07 piksi local0.warn router[987]: send_impl: sendmsg returned EAGAIN, disconnecting and dropping 163993 queued bytes (path: ipc:///var/run/sockets/internal.pub, node: 0x4d400) (endpoint.c:633)
Dec 18 08:16:08 piksi local0.warn resource_monitor[2644]: sbp_rx: reader disconnected, reconnecting...
Dec 18 08:16:10 piksi local0.warn router[987]: send_impl: sendmsg returned EAGAIN, disconnecting and dropping 164222 queued bytes (path: ipc:///var/run/sockets/internal.pub, node: 0x4d400) (endpoint.c:633)
Dec 18 08:16:10 piksi local0.warn resource_monitor[2644]: sbp_rx: reader disconnected, reconnecting...
Dec 18 08:16:11 piksi local0.warn router[987]: send_impl: sendmsg returned EAGAIN, disconnecting and dropping 164266 queued bytes (path: ipc:///var/run/sockets/internal.pub, node: 0x4d400) (endpoint.c:633)
Dec 18 08:16:12 piksi local0.warn resource_monitor[2644]: sbp_rx: reader disconnected, reconnecting...
Dec 18 08:16:15 piksi local0.warn router[987]: send_impl: sendmsg returned EAGAIN, disconnecting and dropping 163868 queued bytes (path: ipc:///var/run/sockets/internal.pub, node: 0x4d400) (endpoint.c:633)
Dec 18 08:16:15 piksi local0.warn resource_monitor[2644]: sbp_rx: reader disconnected, reconnecting...
```

Will need to further investigate how to mitigate this...

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v2.2.9-8-gfce270d3
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=816605)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.2.9-8-gfce270d3&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.2.9-8-gfce270d3&groupby_key=firmware&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v2.2.9-8-gfce270d3
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.2.9-8-gfce270d3)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.2.9-8-gfce270d3/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.2.9-8-gfce270d3

**Comment by silverjam** [comment]
## Verifications

+ Verified that resource_monitor with it's own socket does not get behind on receiving SBP data

+ Verified that resource_monitor messages are still seen on an external port when the daemon is enabled

+ Verified that interval setting can still be modified

**Comment by silverjam** [comment]
Not waiting for Travis, additional changes from last build were config files.

**Pull request merged**